File: tests/property/python/areas/hyperelliptic.py
"""Live-SageMath oracle for the ported hyperelliptic-curve surface."""

from sage.all import *


def _poly(K, coeffs):
    R = PolynomialRing(K, 'x')
    return R([K(c) for c in coeffs])


def _ints(xs):
    return ','.join(str(ZZ(x)) for x in xs)


def hyp_summary(p, f_coeffs, h_coeffs, extensions):
    K = GF(ZZ(p))
    f = _poly(K, f_coeffs)
    h = _poly(K, h_coeffs) if h_coeffs else 0
    H = HyperellipticCurve(f, h)
    counts = H.count_points(int(extensions))
    frob = H.frobenius_polynomial()
    return 'g=%s counts=%s frob=%s jac=%s' % (
        H.genus(),
        _ints(counts),
        _ints(frob.list()),
        frob(1),
    )


def hyp_cartier(p, f_coeffs):
    K = GF(ZZ(p))
    H = HyperellipticCurve(_poly(K, f_coeffs))
    cartier = ';'.join(_ints(row) for row in H.Cartier_matrix().rows())
    hasse_witt = ';'.join(_ints(row) for row in H.Hasse_Witt().rows())
    return 'C=%s HW=%s a=%s p=%s' % (
        cartier,
        hasse_witt,
        H.a_number(),
        H.p_rank(),
    )


FUNCTIONS = {
    'hyp_summary': hyp_summary,
    'hyp_cartier': hyp_cartier,
}

File: tests/property/typescript/areas/hyperelliptic.ts
/**
 * sagemath-ts side of the live hyperelliptic-curve differential area.
 */

import { GF } from '../../../../packages/sagemath-ts/src/rings/finite_rings/index.js';
import type { FiniteFieldElement } from '../../../../packages/sagemath-ts/src/rings/finite_rings/index.js';
import { PolynomialRing } from '../../../../packages/sagemath-ts/src/rings/polynomial/index.js';
import { HyperellipticCurve } from '../../../../packages/sagemath-ts/src/schemes/hyperelliptic_curves/index.js';

type AnyCurve = {
  genus(): number;
  count_points(n: number): bigint[];
  frobenius_polynomial(): bigint[];
  jacobian(): { cardinality(): bigint };
  Cartier_matrix(): FiniteFieldElement[][];
  Hasse_Witt(): FiniteFieldElement[][];
  a_number(): number;
  p_rank(): number;
};

const ints = (xs: Array<bigint | FiniteFieldElement>): string =>
  xs.map((x) => (typeof x === 'bigint' ? x : x.value).toString()).join(',');

function poly(p: bigint, coeffs: bigint[]) {
  const K = GF(p);
  const R = new PolynomialRing<FiniteFieldElement>(K, 'x');
  return R.__call__(coeffs.map((c) => K.__call__(c)));
}

function hyp_summary(p: bigint, fCoeffs: bigint[], hCoeffs: bigint[], extensions: bigint): string {
  const f = poly(p, fCoeffs);
  const h = hCoeffs.length === 0 ? null : poly(p, hCoeffs);
  const H = HyperellipticCurve(f, h) as unknown as AnyCurve;
  return `g=${H.genus()} counts=${ints(H.count_points(Number(extensions)))} frob=${ints(
    H.frobenius_polynomial()
  )} jac=${H.jacobian().cardinality()}`;
}

function hyp_cartier(p: bigint, fCoeffs: bigint[]): string {
  const H = HyperellipticCurve(poly(p, fCoeffs)) as unknown as AnyCurve;
  const C = H.Cartier_matrix().map(ints).join(';');
  const HW = H.Hasse_Witt().map(ints).join(';');
  return `C=${C} HW=${HW} a=${H.a_number()} p=${H.p_rank()}`;
}

export const functions = {
  hyp_summary,
  hyp_cartier,
};
